staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -572,6 +572,19 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(
 		})
 	}
 
+	if len(instanceMeta.AdditionalLabels) > 0 {
+		klog.V(2).Infof("Adding additional node label(s) from cloud provider: %v", instanceMeta.AdditionalLabels)
+		nodeModifiers = append(nodeModifiers, func(n *v1.Node) {
+			if n.Labels == nil {
+				n.Labels = map[string]string{}
+			}
+
+			for k, v := range instanceMeta.AdditionalLabels {
+				n.Labels[k] = v
```

Comment:
Do we care if the key is already present or not?

Reply:
Hmm, good question. Probably not? The issue would be overriding someone else's value, but cloud provider should only be setting these labels if they're authoritative, IMO.

Reply:
In https://github.com/kubernetes/kubernetes/pull/123223#issuecomment-1937086154 I suggest that we allow it, and we also record an event once per node if we detect this overwriting.

I guess we could track a metric (counter) instead; either's fine. My point is that a cluster admin might want a way to tell why the label doesn't look right, and to rule out this new mechanism as the culprit.

Reply:
Ok. I've added a log for each property getting overridden, and I've added an event in this case. The log itself could be enough to assist cluster admins, but the event would bring even more visibility. I don't think a metric is the right option because it would probably be hard to find if the operator was trying to find the culprit, though I could be wrong on that.

That being said, this case should be the exception and not the rule.

Reply:
I'm curious what folks think about appropriate visibility here. I think that the log alone is *probably* enough...

Reply:
I assume an event is cheap, so if there aren't any perf concerns I think it's worth adding.

Reply:
I'm not in favor of overridden as this can create hot loop situations where controllers set labels , so the controller set label foo and this changes it , and then the control changes it again ... is not a good UX and it will cause problems hard to detect, all the situations I've seen like these, events were not great to pinpoint these problems

Reply:
> I'm not in favor of overridden as this can create hot loop situations

Ya. That's a good point. From a user experience perspective, I think it's going to be confusing whether you override or ignore. The ideal solution would be to allow cloud provider implementations to reserve label namespaces to avoid nodes and other callers from using them specifically, but providing a mechanism for that wouldn't be practical because it wouldn't be backwards compatible.

A similar point on the loop, if we were to emit event, we'd need to ensure that we don't emit an event EVERY time that the controller is run, so I don't think the event makes sense if we didn't override the value and if we did override it, we should only emit an emit when the value changes, which I'm not doing at the moment.

Anyway, I think your argument is persuasive that overriding is probably not the right behavior.